### PR TITLE
Update to PQMass

### PR DESCRIPTION
**Pull request by SammyS15**
I have done the following
1) Updated the README to include more detail about PQMass as well as examples (in distribution and out of distribution) as well as advanced options
2) Made PQMass GPU compatible that have the same results as the cpu (default) option
3) Renamed "whiten" to "z_score_norm" as whiten doesn't tell the user what they are working with
4) Updated notebooks to include device parameter to showcase that it could be 'cuda' if the user has access to GPU

To Do Later
1) More encompassing test cases

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,63 +6,152 @@
...
+
+[PQMass](https://arxiv.org/abs/2402.04355) is a new sample-based method for evaluating the quality of generative models as well assessing distribution shifts.
 
 ## Install
 
-Just do:
+To install PQMass, run the following:
 
-```
+```python
```
Comment (ConnorStoneAstro) [inline README.md:17]: I think it would be `bash` not `python` since you dont run pip in python

File: README.md [hunk 2]
```diff
@@ -6,63 +6,152 @@
 ![PyPI - Downloads](https://img.shields.io/pypi/dm/pqm)
 [![arXiv](https://img.shields.io/badge/arXiv-2402.04355-b31b1b.svg)](https://arxiv.org/abs/2402.04355)
 
-Implementation of the PQMass two sample test from Lemos et al. 2024 [here](https://arxiv.org/abs/2402.04355)
+<!-- Implementation of the PQMass two sample test from Lemos et al. 2024 [here](https://arxiv.org/abs/2402.04355) -->
```
Comment (ConnorStoneAstro) [inline README.md:9]: I would remove rather than comment this out

File: README.md [hunk 3]
```diff
@@ -6,63 +6,152 @@
...
+
+[PQMass](https://arxiv.org/abs/2402.04355) is a new sample-based method for evaluating the quality of generative models as well assessing distribution shifts.
```
Comment (ConnorStoneAstro) [inline README.md:11]: The point about generative models is good, but you might want to add some more to mention that the method works for comparing any two samples to see if they come from the same distribution

File: README.md [hunk 4]
```diff
@@ -6,63 +6,152 @@
...
+you get a very small value (i.e., 1e-6), then you have failed the null
+hypothesis test, and the two samples are not drawn from the same distribution.
+If you get values approximately equal to 1 every time then that suggests
+potential duplication of samples between `x` and `y`.
+
+PQMass can work for any two datasets as it measures the distribution shift between the $x$ and $y$, which we show below.
+
+## Example
+
+We are using 100 regions. Thus, the DoF is 99, our expected $\chi^2$ peak of the distribution is 97, the median is 99, and the standard deviation should be 14.07. With this in mind, we set up our example. For the p-value, we expect to be between 0 and 1 and a significantly small p-value (e.g., $< 0.05$ or $< 0.01$) would mean we reject the null hypothesis and thus $x$ and $y$ do not come from the same distribution.
```
Comment (ConnorStoneAstro) [inline README.md:53]: the mean is 99. The median may be something else

File: src/pqm/pqm.py [hunk 5]
```diff
@@ -1,35 +1,119 @@
...
+        # Compute p-value using the survival function (1 - CDF)
+        p_value = torch.special.gammaincc(dof / 2, chi2_stat / 2).item()
+
+        return chi2_stat, p_value, dof, E
+
+
 def _pqm_test(
-    x_samples: np.ndarray,
-    y_samples: np.ndarray,
+    x_samples: torch.Tensor,
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:110]: We should definitely keep numpy input as an option since it is less aggressive than pytorch about requiring everything to be arrays. Numpy is a good common ground say for people who use JAX that still want to use PQM. I'm sure JAX has a to_numpy function, but not a to_torch option.

File: src/pqm/pqm.py [hunk 6]
```diff
@@ -38,14 +122,22 @@ def _pqm_test(
...
     num_refs : int
-        Number of reference samples to use.
-    whiten : bool
-        If True, whiten the samples by subtracting the mean and dividing by the
-        standard deviation.
+        Number of reference samples to use. These samples will be drawn from
+        x_samples, y_samples, and/or a Gaussian distribution, see the note
+        below.
+    re_tessellation : Optional[int]
+        Number of times pqm_pvalue is called, re-tesselating the space. No
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:135]: `_pqm_test` not `pqm_pvalue`

File: src/pqm/pqm.py [hunk 7]
```diff
@@ -58,6 +150,21 @@ def _pqm_test(
         support of the reference samples if pathological behavior is expected.
         Default: 0.0 no gaussian samples.
 
+    Note
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:153]: it looks like this note is duplicated?

File: src/pqm/pqm.py [hunk 8]
```diff
@@ -86,56 +193,81 @@ def _pqm_test(
...
+    probs = torch.tensor(
+        [
+            x_frac * (1.0 - gauss_frac),
+            (1.0 - x_frac) * (1.0 - gauss_frac),
+            gauss_frac,
+        ],
+        device=device,
+    )
+
+    counts = Multinomial(total_count=num_refs, probs=probs).sample()
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:214]: I think the numpy version is easier to use in one line here, rather than using pytorch then converting tolist in the end

File: src/pqm/pqm.py [hunk 9]
```diff
@@ -86,56 +193,81 @@ def _pqm_test(
...
+    assert (Nx + Ny + Ng) == num_refs, (
+        f"Something went wrong. Nx={Nx}, Ny={Ny}, Ng={Ng} should sum to num_refs={num_refs}"
     )
-    assert (Nx + Ny + Ng) == num_refs, f"Something went wrong. Nx={Nx}, Ny={Ny}, Ng={Ng} should sum to num_refs={num_refs}"  # fmt: skip
 
     # Collect reference samples from x_samples
-    xrefs = np.random.choice(nx, Nx, replace=False)
-    xrefs, x_samples = x_samples[xrefs], np.delete(x_samples, xrefs, axis=0)
+    x_indices = torch.randperm(nx, device=device)
+    if Nx > nx:
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:223]: this error should be up where `Nx` is determined

File: src/pqm/pqm.py [hunk 10]
```diff
@@ -86,56 +193,81 @@ def _pqm_test(
...
+    x_samples_indices = x_indices[Nx:]
+
+    xrefs = x_samples[xrefs_indices]
+    x_samples = x_samples[x_samples_indices]
 
     # Collect reference samples from y_samples
-    yrefs = np.random.choice(ny, Ny, replace=False)
-    yrefs, y_samples = y_samples[yrefs], np.delete(y_samples, yrefs, axis=0)
+    y_indices = torch.randperm(ny, device=device)
+    if Ny > ny:
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:233]: this error should be up where `Ny` is determined

File: src/pqm/pqm.py [hunk 11]
```diff
@@ -86,56 +193,81 @@ def _pqm_test(
...
+    Nx, Ny, Ng = counts.tolist()
+    assert (Nx + Ny + Ng) == num_refs, (
+        f"Something went wrong. Nx={Nx}, Ny={Ny}, Ng={Ng} should sum to num_refs={num_refs}"
     )
-    assert (Nx + Ny + Ng) == num_refs, f"Something went wrong. Nx={Nx}, Ny={Ny}, Ng={Ng} should sum to num_refs={num_refs}"  # fmt: skip
 
     # Collect reference samples from x_samples
-    xrefs = np.random.choice(nx, Nx, replace=False)
-    xrefs, x_samples = x_samples[xrefs], np.delete(x_samples, xrefs, axis=0)
+    x_indices = torch.randperm(nx, device=device)
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:222]: again, I think we need to keep both numpy and torch versions. Maybe we need a `_pqm_test_numpy` and `_ppqm_test_torch`?

File: src/pqm/pqm.py [hunk 12]
```diff
@@ -215,37 +350,49 @@ def pqm_pvalue(
...
+        Device to use for computation. Default: 'cpu'. If 'cuda' is selected,
 
     Returns
     -------
     float or list
         pvalue(s). Null hypothesis that both samples are drawn from the same
         distribution.
     """
+    # Move samples to torch tensors on the selected device
+    x_samples = torch.tensor(x_samples, device=device)
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:363]: for this you would use `torch.as_tensor` in case the samples are already a tensor

File: src/pqm/pqm.py [hunk 13]
```diff
@@ -312,26 +461,31 @@ def pqm_chi2(
     float or list
         chi2 statistic(s).
     """
+    # Move samples to torch tensors on the selected device
+    x_samples = torch.tensor(x_samples, device=device)
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:465]: as_tensor

File: src/pqm/test_gaussian.py [hunk 14]
```diff
@@ -8,6 +8,6 @@ def test():
         y_samples = np.random.normal(size=(500, 50))
         x_samples = np.random.normal(size=(250, 50))
 
-        new.append(pqm_pvalue(x_samples, y_samples))
+        new.append(pqm_pvalue(x_samples, y_samples, device = 'cpu'))
```
Comment (ConnorStoneAstro) [inline src/pqm/test_gaussian.py:11]: device defaults to cpu, so this isn't needed

**Review by ConnorStoneAstro — CHANGES_REQUESTED** [review]
Lots of good stuff in here! But I think we need to keep the numpy option, when running on CPU the version with the KDTree is much faster, which is the main use case.

File: src/pqm/pqm.py [hunk 15]
```diff
@@ -1,35 +1,119 @@
...
+    tuple
+        chi2_stat, p_value, dof, expected
+    """
+    if device == 'cpu':
+        counts_np = counts.cpu().numpy()
+        chi2_stat, p_value, dof, expected = chi2_contingency(counts_np)
+        chi2_stat = torch.tensor(chi2_stat, device=device)
+        dof = torch.tensor(dof, device=device)
+        return chi2_stat, p_value, dof, expected
+    else:
```
[the remark this answers is not in the document]
Comment (SammyS15) [reply src/pqm/pqm.py]: I think leaving the torch version is fine and it has shown to be accurate so I will leave it as is

File: README.md [hunk 16]
```diff
@@ -6,63 +6,158 @@
...
+print(np.mean(chi2_stat), np.std(chi2_stat)) # 98.51, 11.334
 
 # To get pvalues from PQMass
-pvalues = pqm_pvalue(x_sample, y_sample, num_refs = 100, re_tessellation = 50)
-print(np.mean(pvalues), np.std(pvalues))
+pvalues = pqm_pvalue(p, q, re_tessellation = 1000)
+print(np.mean(pvalues), np.std(pvalues)) # 0.50, 0.26
+```
+
+We see that both $\chi_{PQM}^2$ and $\text{p-value}(\chi_{PQM}^2)$ follow the expected $\chi^2$ indicatiing that both $x$ and $y$ come from the same underlying distribution.
```
Comment (ConnorStoneAstro) [inline README.md:73]:
```suggestion
We see that both $\chi_{PQM}^2$ and $\text{p-value}_{PQM}$ follow the expected $\chi^2$ indicating that both $x$ and $y$ come from the same underlying distribution.
```

File: README.md [hunk 17]
```diff
@@ -6,63 +6,158 @@
...
-distribution. The degrees of freedom (DoF) will equal `DoF = num_refs - 1` The
-peak of this distribution will be at `DoF - 2`, the mean will equal `DoF`, and
-the standard deviation will be `sqrt(2 * DoF)`. If your chi^2 values are too
-high (`chi^2 / DoF > 1`), it suggests that the samples are out of distribution.
-Conversely, if the values are too low (`chi^2 / DoF < 1`), it indicates
-potential duplication of samples between `x_samples` and `y_samples` (i.e.
-memorization for generative models).
+Thus, PQMass can be used to identify if any two distributions come from the same underlying distributions if enough samples are given. We encourage users to look through the paper to see the varying experiments and use cases for PQMass!
+
+## How to Intrept Result
```
Comment (ConnorStoneAstro) [inline README.md:97]: Maybe this could be part of an FAQ? It's not really about interpreting the results generally, but more about a specific case that can go wrong

File: src/pqm/pqm.py [hunk 18]
```diff
@@ -1,51 +1,402 @@
...
+    num_refs : int
+        Number of reference samples to use.
+
+    Returns
+    -------
+    tuple
+        Results from the PyTorch implementation of chi2_contingency.
+    """
+
+    # Compute distances and find nearest references
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:330]: Should rearrange either this one or the numpy one so that they are organized the same way. This does all x then all y, whereas the numpy one does both cdist, both argmin, both bincount.

File: src/pqm/pqm.py [hunk 19]
```diff
@@ -74,98 +427,94 @@ def _pqm_test(
...
         )
-    if whiten:
+    
+    # Z-score normalization
+    if z_score_norm:
         mean, std = _mean_std(x_samples, y_samples)
-        y_samples = (y_samples - mean) / std
-        x_samples = (x_samples - mean) / std
-
+        if is_numpy:
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:456]: You don't need the if statement since both numpy and torch versions are the same

File: src/pqm/pqm.py [hunk 20]
```diff
@@ -74,98 +427,94 @@ def _pqm_test(
...
-    return chi2_contingency(np.stack([counts_x, counts_y]))
-
+    
+    # Sampling reference indices
+    if is_numpy:
+        refs = _sample_reference_indices_numpy(Nx, nx, Ny, ny, Ng, x_samples, y_samples)
+    elif is_torch:
+        refs = _sample_reference_indices_torch(Nx, nx, Ny, ny, Ng, x_samples, y_samples, device)
+    
+    # Update num_refs in case Gaussian samples were added
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:499]: the total of `Nx+Ny+Ng` is always `num_refs` no need for `current_num_refs`

File: src/pqm/pqm.py [hunk 21]
```diff
@@ -185,10 +534,10 @@ def pqm_pvalue(
         x_samples, y_samples, and/or a Gaussian distribution, see the note
         below.
     re_tessellation : Optional[int]
-        Number of times pqm_pvalue is called, re-tesselating the space. No
+        Number of times _pqm_test is called, re-tesselating the space. No
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:537]: I don't think the user needs to know that we've organized the code to use a common `_pqm_test` function, just saying that the code is rerun a bunch of times is enough

File: src/pqm/pqm.py [hunk 22]
```diff
@@ -216,36 +567,58 @@ def pqm_pvalue(
...
+
     Returns
     -------
     float or list
         pvalue(s). Null hypothesis that both samples are drawn from the same
         distribution.
     """
+    # Check the device and convert to the respective type (Numpy or Torch) and call their respective _pqm_test function
+
+    if device.type == 'cpu':
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:579]: This won't work if someone enters `device="cpu"` since str doesn't have the attribute `type`

File: src/pqm/pqm.py [hunk 23]
```diff
@@ -216,36 +567,58 @@ def pqm_pvalue(
...
-    _, pvalue, _, _ = _pqm_test(x_samples, y_samples, num_refs, whiten, x_frac, gauss_frac)
-    return pvalue
-
+    
+    _, p_value, _, _ = _pqm_test(
+        x_samples, y_samples, num_refs, z_score_norm, x_frac, gauss_frac, device
+    )
+    
+    # Return p-value as a float
+    return p_value if isinstance(p_value, float) else float(p_value)
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:611]: I think `p_value` will always be a float, so this check isn't needed

File: src/pqm/pqm.py [hunk 24]
```diff
@@ -254,21 +627,21 @@ def pqm_chi2(
...
+    y_samples : np.ndarray or torch.Tensor
         Samples from the second distribution. Must have shape (M, *D) M is the
         number of y samples, and D is the dimensionality of the samples.
     num_refs : int
         Number of reference samples to use. These samples will be drawn from
         x_samples, y_samples, and/or a Gaussian distribution, see the note
         below.
     re_tessellation : Optional[int]
-        Number of times pqm_pvalue is called, re-tesselating the space. No
+        Number of times _pqm_test is called, re-tesselating the space. No
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:641]: same here

File: src/pqm/pqm.py [hunk 25]
```diff
@@ -312,26 +687,43 @@ def pqm_chi2(
     float or list
         chi2 statistic(s).
     """
+
+    # Check the device and convert to the respective type (Numpy or Torch) and call their respective _pqm_test function
+    if device.type == 'cpu':
```
Comment (ConnorStoneAstro) [inline src/pqm/pqm.py:692]: same here

**Review by ConnorStoneAstro — CHANGES_REQUESTED** [review]
Some small comments, easy to fix

File: README.md (hunk 16 above)
Comment (SammyS15) [reply README.md:73]: I am leaving $\text{p-value}(\chi_{PQM}^2)$ as this is how it is noted in Equation 12 in the paper

File: src/pqm/pqm.py (hunk 23 above)
Comment (SammyS15) [reply src/pqm/pqm.py:611]: I am inclined to keep this just in case, I don't see the reason to remove it

File: src/pqm/pqm.py (hunk 25 above)
Comment (SammyS15) [reply src/pqm/pqm.py:692]: Good catch. I have updated it to be if device.type == 'cpu' or device == 'cpu':

**Review by ConnorStoneAstro — APPROVED** [review]
